### PR TITLE
Validate rule before kicking off workflows or creating event source mappings

### PR DESCRIPTION
**Pull request by laurenfrederick**
Addresses [CUMULUS-1854: Develop amazing new feature](https://bugs.earthdata.nasa.gov/browse/CUMULUS-1854)

## PR Checklist

- [ ] Update CHANGELOG
- [ ] Unit tests
- [ ] Adhoc testing
- [ ] Integration tests

### CODE REVIEW

File: packages/api/tests/models/test-rules-model.js [hunk 1]
```diff
@@ -358,6 +362,23 @@ test.serial('Creating a kinesis rule where an event source mapping already exist
   }
 });
 
+test.serial('creating an invalid kinesis type rule does not add event mappings', async (t) => {
+  const { kinesisRule } = t.context;
+
+  const newKinesisRule = cloneDeep(kinesisRule);
```
Comment (markdboyd) [inline packages/api/tests/models/test-rules-model.js:368]: Nit: you shouldn't need to `cloneDeep` because `t.context.kinesisRule` is reset in `beforeEach` for every test

File: packages/api/models/rules.js [hunk 2]
```diff
@@ -228,6 +228,12 @@ class Rule extends Manager {
       newRuleItem.state = 'ENABLED';
     }
 
+    newRuleItem.createdAt = Date.now();
+    newRuleItem.updatedAt = Date.now();
+
+    // Validate rule before kicking off workflows or adding event source mappings
+    await this.constructor.recordIsValid(newRuleItem, this.schema, this.removeAdditional);
```
Comment (markdboyd) [inline packages/api/models/rules.js:235]: Super, super good change

File: packages/api/tests/models/test-rules-model.js [hunk 3]
```diff
@@ -358,6 +362,23 @@ test.serial('Creating a kinesis rule where an event source mapping already exist
   }
 });
 
+test.serial('creating an invalid kinesis type rule does not add event mappings', async (t) => {
```
Comment (markdboyd) [inline packages/api/tests/models/test-rules-model.js:365]: Do we care about having the same test to ensure SNS triggers aren't created for invalid rules? I could go either way

Comment (laurenfrederick) [reply packages/api/tests/models/test-rules-model.js:365]: My thinking was if we tested one that was good, but what do you think? I mean I could go either way too, so took the easy way out.

File: packages/api/tests/models/test-rules-model.js (hunk 1 above)
Comment (laurenfrederick) [reply packages/api/tests/models/test-rules-model.js:368]: Oh yeah that is a good point. Habit I guess!

File: packages/api/tests/models/test-rules-model.js (hunk 3 above)
Comment (markdboyd) [reply packages/api/tests/models/test-rules-model.js:365]: I guess I incline towards being thorough but it does seem repetitive. Alternatively, if all the code in the rules model from lines 237 - 264 was in a separate (private) method, then we could just test using sinon that the method is never called when you try to create an invalid rule.

More importantly, we probably also want to explicitly check that the rule is valid in `rule.update` before updating rule triggers?

Comment (laurenfrederick) [reply packages/api/tests/models/test-rules-model.js]: Yep, like to all of that. Took your suggestions.

File: packages/api/models/rules.js [hunk 4]
```diff
@@ -134,50 +134,61 @@ class Rule extends Manager {
...
-      await this.addRule(updatedRuleItem, payload);
-      break;
-    }
-    case 'kinesis':
-      if (valueUpdated) {
-        await this.deleteKinesisEventSources(updatedRuleItem);
-        const updatedRuleItemArns = await this.addKinesisEventSources(updatedRuleItem);
-        updatedRuleItem = this.updateKinesisRuleArns(updatedRuleItem,
-          updatedRuleItemArns);
+      case 'scheduled': {
```
Comment (markdboyd) [inline packages/api/models/rules.js:154]: When I check this out locally, ESlint is complaining about the spaces, though I don't know why

File: packages/api/tests/models/test-rules-model.js [hunk 5]
```diff
@@ -31,6 +31,18 @@ async function getKinesisEventMappings() {
...
 }
 
+async function deleteKinesisEventSourceMappings() {
+  const eventMappings = await getKinesisEventMappings();
+
+  if(!eventMappings) {
+    return Promise.resolve();
+  }
+
+  const allEventMappings = eventMappings[0].EventSourceMappings.concat(eventMappings[1].EventSourceMappings);
```
Comment (markdboyd) [inline packages/api/tests/models/test-rules-model.js:41]: Eslint is complaining about the length of this line

File: packages/api/tests/models/test-rules-model.js [hunk 6]
```diff
@@ -31,6 +31,18 @@ async function getKinesisEventMappings() {
...
+
+  return Promise.all(allEventMappings.map((e) => awsServices.lambda().deleteEventSourceMapping({ UUID: e.UUID }).promise()));
```
Comment (markdboyd) [inline packages/api/tests/models/test-rules-model.js:43]: Eslint is complaining about the length of this line

File: packages/api/models/rules.js [hunk 7]
```diff
@@ -228,39 +239,51 @@ class Rule extends Manager {
...
     // save
     return super.create(newRuleItem);
   }
 
+  async createRuleTrigger(ruleItem) {
+    let newRuleItem = cloneDeep(ruleItem);
+
+    const payload = await Rule.buildPayload(newRuleItem);
+    switch (newRuleItem.rule.type) {
+      case 'onetime': {
```
Comment (markdboyd) [inline packages/api/models/rules.js:259]: Weirdly, ESlint seems to think that case statements should not be indented more than the switch statement.

File: packages/api/tests/models/test-rules-model.js [hunk 8]
```diff
@@ -149,10 +161,27 @@ test('Creating a rule with an invalid type throws an error', async (t) => {
...
   await t.throwsAsync(
     () => rulesModel.create(ruleItem),
-    { message: 'Rule type \'invalid\' not supported.' }
+    { name: 'SchemaValidationError' }
   );
 });
 
+test.serial('Creating an invalid rule does not create workflow triggers', async (t) => {
+  const { onetimeRule } = t.context;
+  const ruleItem = cloneDeep(onetimeRule);
```
Comment (markdboyd) [inline packages/api/tests/models/test-rules-model.js:170]: Nit: You don't need to cloneDeep

File: packages/api/tests/models/test-rules-model.js [hunk 9]
```diff
@@ -149,10 +161,27 @@ test('Creating a rule with an invalid type throws an error', async (t) => {
...
+
+  const createTriggerStub = sinon.stub(models.Rule.prototype, 'createRuleTrigger').resolves(ruleItem);
+
+  await t.throwsAsync(
+    () => rulesModel.create(ruleItem),
+    { name: 'SchemaValidationError' }
+  );
+
+  t.true(createTriggerStub.notCalled);
+  createTriggerStub.restore();
```
Comment (markdboyd) [inline packages/api/tests/models/test-rules-model.js:182]:
```suggestion
  const createTriggerStub = sinon.stub(models.Rule.prototype, 'createRuleTrigger').resolves(ruleItem);

  try {
    await t.throwsAsync(
      () => rulesModel.create(ruleItem),
      { name: 'SchemaValidationError' }
    );
    t.true(createTriggerStub.notCalled);
  } finally {
    createTriggerStub.restore();
  }
```

We want to make sure that this stub gets restored even if the test fails on `throwsAsync`, so that the stub isn't still active when other tests run. We also want to do this for the `update` test

File: packages/api/tests/models/test-rules-model.js [hunk 10]
```diff
@@ -311,6 +362,7 @@ test.serial('create a kinesis type rule, using existing event source mappings',
 
   await rulesModel.delete(rule);
   await rulesModel.delete(newRule);
+  await deleteKinesisEventSourceMappings();
```
Comment (markdboyd) [inline packages/api/tests/models/test-rules-model.js:365]: I don't doubt that it's necessary, but why did you add these to most of the tests? To ensure test isolation?

**Comment by markdboyd** [comment]
@laurenfrederick Looks great! Just a few small cleanup comments

File: packages/api/tests/models/test-rules-model.js (hunk 10 above)
Comment (laurenfrederick) [reply packages/api/tests/models/test-rules-model.js:367]: Yes, when I was writing the test the other way this was needed and I figured there's no reason to go delete it since it's actually a good change to isolate the tests and may help someone down the road.

File: packages/api/tests/models/test-rules-model.js (hunk 8 above)
Comment (chuckwondo) [reply packages/api/tests/models/test-rules-model.js:178]: I'm curious about this. Why is `cloneDeep` not necessary?

I see that ava does attempt to isolate data between tests (https://github.com/avajs/ava/blob/master/docs/01-writing-tests.md#test-context).

However, it's not clear that a deep clone is performed on the context. Specifically:

> Context created in .before() hooks is cloned before it is passed to .beforeEach() hooks and / or tests. The .after() and .after.always() hooks receive the original context value.
>
> For .beforeEach(), .afterEach() and .afterEach.always() hooks the context is not shared between different tests, allowing you to set up data such that it will not leak to other tests.

The first sentence links to lodash's shallow clone (see the ava docs at the link above), but the second sentence seems to _imply_ that some sort of deep clone is happening, but it is not clear as to what actually happens.

Comment (markdboyd) [reply packages/api/tests/models/test-rules-model.js:177]: `t.context.onetimeRule` is created in `beforeEach`. And as the docs you posted say, context set up in beforeEach is not shared between tests. So there's no need to clone deep because `t.context.onetimeRule` will already be a new object in each test

Comment (chuckwondo) [reply packages/api/tests/models/test-rules-model.js:177]: Oh, I see. That makes sense now. Thanks!
